akka-http-core/src/main/scala/akka/http/impl/engine/http2/OutgoingConnectionBuilderImpl.scala

```diff
@@ -0,0 +1,156 @@
+import akka.http.scaladsl.model.HttpRequest
+import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.TLSClosing
+import akka.stream.impl.io.TlsUtils
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.TLS
+import akka.stream.scaladsl.TLSPlacebo
+import javax.net.ssl.SSLEngine
+import akka.http.javadsl
+import akka.http.javadsl.model
+import akka.http.javadsl.{ OutgoingConnectionBuilder => JOutgoingConnectionBuilder }
+import akka.http.scaladsl.ConnectionContext
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.HttpsConnectionContext
+import akka.http.scaladsl.OutgoingConnectionBuilder
+import akka.stream.javadsl.{ Flow => JFlow }
+
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object OutgoingConnectionBuilderImpl {
+
+  def apply(host: String, system: ClassicActorSystemProvider): OutgoingConnectionBuilder =
+    Impl(
+      host,
+      None,
+      clientConnectionSettings = ClientConnectionSettings(system),
+      connectionContext = None,
+      log = system.classicSystem.log,
+      system = system,
+      usingHttp2 = false
+    )
+
+  private final case class Impl(
+    host:                     String,
+    port:                     Option[Int],
+    clientConnectionSettings: ClientConnectionSettings,
+    connectionContext:        Option[HttpsConnectionContext],
+    log:                      LoggingAdapter,
+    system:                   ClassicActorSystemProvider,
+    usingHttp2:               Boolean) extends OutgoingConnectionBuilder {
+
+    override def toHost(host: String): OutgoingConnectionBuilder = copy(host = host)
+
+    override def toPort(port: Int): OutgoingConnectionBuilder = copy(port = Some(port))
+
+    override def withCustomHttpsConnectionContext(httpsConnectionContext: HttpsConnectionContext): OutgoingConnectionBuilder = copy(connectionContext = Some(httpsConnectionContext))
+
+    override def withClientConnectionSettings(settings: ClientConnectionSettings): OutgoingConnectionBuilder = copy(clientConnectionSettings = settings)
+
+    override def logTo(logger: LoggingAdapter): OutgoingConnectionBuilder = copy(log = logger)
+
+    override def http(): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]] = {
+      // http/1.1 plaintext
+      Http(system).outgoingConnectionUsingContext(host, port.getOrElse(80), ConnectionContext.noEncryption(), clientConnectionSettings, log)
+    }
+
+    override def https(): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]] = {
+      // http/1.1 tls
+      Http(system).outgoingConnectionHttps(host, port.getOrElse(443), connectionContext.getOrElse(Http(system).defaultClientHttpsContext), None, clientConnectionSettings, log)
+    }
+
+    override def http2(): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]] = {
+      // http/2 tls
+      val port = this.port.getOrElse(443)
```

Comment:
All of this should stay / be implemented in `Http2`.

Reply:
Alright, so both http2 variations go in there as internal api and just delegate from here?

Reply:
Yep.